src/Dialect/ONNX/Transforms/Decompose.cpp

```diff
@@ -934,15 +935,17 @@ class ReplaceCastLikeByCastPattern : public OpRewritePattern<ONNXCastLikeOp> {
     Location loc = castLikeOp.getLoc();
 
     Value input = castLikeOp.getInput();
+    Value output = castLikeOp.getOutput();
     Value target = castLikeOp.getTargetType();
     IntegerAttr saturate = castLikeOp.getSaturateAttr();
 
     // The output type will be the same as the target_type or the second input
-    Type outputType = target.getType().cast<ShapedType>().getElementType();
+    Type targetType = target.getType().cast<ShapedType>().getElementType();
 
     // Replace
-    Value res = onnx_mlir::OnnxBuilder(rewriter, loc)
-                    .cast(input, saturate, TypeAttr::get(outputType));
+    Value res =
+        onnx_mlir::OnnxBuilder(rewriter, loc)
+            .cast(output.getType(), input, saturate, TypeAttr::get(targetType));
```

Comment:
I think we can use an unknown result type, then DialectBuilder will infer shape when constructing the op.
```
Type resultType = UnrankedTensorType::get(targetType);
Value res = onnx_mlir::OnnxBuilder(rewriter, loc).cast(resultType, input, saturate, TypeAttr::get(targetType));
 ```

Reply:
Updated. I checked if output type has rank, and if not, this code is used.